File: rs/ll-open/fs/tests/smoke_ffi.c
/*
 * C smoke test for leyline-fs FFI.
 *
 * This test links against libleyline_fs and exercises the full handle lifecycle:
 *   leyline_open → get_node / list_children / lookup_child / read_content → leyline_close
 *
 * Build & run (from rs/):
 *   cargo build
 *   cc -o /tmp/smoke_ffi crates/fs/tests/smoke_ffi.c \
 *      -I crates/fs/include \
 *      -L target/debug -lleyline_fs \
 *      -framework Security -framework CoreFoundation \
 *      -lSystem -lresolv -lc
 *   /tmp/smoke_ffi /path/to/control_file
 *
 * For automated testing, the Rust integration test (smoke_ffi_integration)
 * creates a temp arena and invokes this binary.
 */

#include <stdio.h>
#include <string.h>
#include "leyline_fs.h"

#define BUF_SIZE 4096

static int failures = 0;

#define CHECK(cond, msg) do { \
    if (!(cond)) { \
        fprintf(stderr, "FAIL: %s\n", msg); \
        failures++; \
    } else { \
        printf("  ok: %s\n", msg); \
    } \
} while (0)

int main(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <control-path>\n", argv[0]);
        return 1;
    }

    printf("=== leyline-fs C smoke test ===\n");

    /* 1. Open */
    struct LeylineCtx *ctx = leyline_open(argv[1]);
    CHECK(ctx != NULL, "leyline_open returned non-null handle");
    if (!ctx) return 1;

    uint8_t buf[BUF_SIZE];
    int32_t n;

    /* 2. get_node — root (empty string) */
    n = leyline_get_node(ctx, "", buf, BUF_SIZE);
    CHECK(n > 0, "get_node root returns data");
    if (n > 0) {
        buf[n] = '\0';
        printf("    root: %s\n", buf);
        CHECK(strstr((char*)buf, "\"kind\":1") != NULL, "root is a directory (kind=1)");
    }

    /* 3. list_children — root should have children */
    n = leyline_list_children(ctx, "", buf, BUF_SIZE);
    CHECK(n > 0, "list_children root returns data");
    if (n > 0) {
        buf[n] = '\0';
        printf("    children: %s\n", buf);
        CHECK(buf[0] == '[', "children is a JSON array");
    }

    /* 4. lookup_child — find "vulns" under root */
    n = leyline_lookup_child(ctx, "", "vulns", buf, BUF_SIZE);
    CHECK(n > 0, "lookup_child('', 'vulns') found");
    if (n > 0) {
        buf[n] = '\0';
        CHECK(strstr((char*)buf, "\"id\":\"vulns\"") != NULL, "child id is 'vulns'");
    }

    /* 5. read_content — leaf node */
    n = leyline_read_content(ctx, "vulns/CVE-1", buf, BUF_SIZE, 0);
    CHECK(n > 0, "read_content returns data");
    if (n > 0) {
        buf[n] = '\0';
        printf("    content: %s\n", buf);
    }

    /* 6. Null handle returns -2 */
    n = leyline_get_node(NULL, "vulns", buf, BUF_SIZE);
    CHECK(n == -2, "null ctx returns -2");

    /* 7. Missing node returns -1 */
    n = leyline_get_node(ctx, "nonexistent", buf, BUF_SIZE);
    CHECK(n == -1, "missing node returns -1");

    /* 8. Close */
    leyline_close(ctx);
    printf("  ok: leyline_close succeeded\n");

    /* 9. Close null is safe */
    leyline_close(NULL);
    printf("  ok: leyline_close(NULL) is safe\n");

    printf("\n=== %s (%d failures) ===\n", failures == 0 ? "PASS" : "FAIL", failures);
    return failures > 0 ? 1 : 0;
}

File: rs/ll-open/fs/tests/smoke_ffi.rs
//! Integration test: compile the C smoke test and run it against a temp arena.
//!
//! Creates a real arena file + control file on disk, populates it with
//! a `nodes` table, then compiles `tests/smoke_ffi.c` against `libleyline_fs`
//! and invokes the resulting binary.

use anyhow::Result;
use leyline_core::{ArenaHeader, Controller};
use leyline_schema::create_schema;
use rusqlite::{Connection, DatabaseName};
use std::process::Command;
use tempfile::TempDir;

/// Build a minimal arena file: [Header][Buf0][Buf1] with a nodes-table DB
/// in the active buffer, plus a control file pointing at it.
fn create_test_arena(dir: &std::path::Path) -> Result<std::path::PathBuf> {
    let source = Connection::open_in_memory()?;
    create_schema(&source)?;
    source.execute_batch(
        "INSERT INTO nodes (id, parent_id, name, kind, size, mtime, record) VALUES ('vulns', '', 'vulns', 1, 0, 1000, NULL);
        INSERT INTO nodes (id, parent_id, name, kind, size, mtime, record) VALUES ('vulns/CVE-1', 'vulns', 'CVE-1', 0, 23, 2000, '{\"severity\":\"critical\"}');
        INSERT INTO nodes (id, parent_id, name, kind, size, mtime, record) VALUES ('vulns/CVE-2', 'vulns', 'CVE-2', 0, 10, 3000, '{\"severity\":\"high\"}');",
    )?;
    let serialized = source.serialize(DatabaseName::Main)?;
    let db_bytes = serialized.as_ref();

    let buf_size = db_bytes.len().max(4096);
    let arena_size = 4096 + buf_size * 2;
    let mut arena = vec![0u8; arena_size];

    // Header: active_buffer = 0
    let header = ArenaHeader {
        magic: ArenaHeader::MAGIC,
        version: ArenaHeader::VERSION,
        active_buffer: 0,
        padding: [0; 2],
        sequence: 1,
    };
    let header_bytes: &[u8] = bytemuck::bytes_of(&header);
    arena[..header_bytes.len()].copy_from_slice(header_bytes);

    // Write DB into buffer 0 (offset = 4096)
    arena[4096..4096 + db_bytes.len()].copy_from_slice(db_bytes);

    let arena_path = dir.join("test.arena");
    std::fs::write(&arena_path, &arena)?;

    // Create control file pointing at the arena
    let ctrl_path = dir.join("test.ctrl");
    let mut ctrl = Controller::open_or_create(&ctrl_path)?;
    ctrl.set_arena(arena_path.to_str().unwrap(), arena_size as u64, 1)?;

    Ok(ctrl_path)
}

#[test]
fn smoke_ffi_c_binary() -> Result<()> {
    let tmp = TempDir::new()?;
    let ctrl_path = create_test_arena(tmp.path())?;

    let crate_dir = std::path::Path::new(env!("CARGO_MANIFEST_DIR"));
    let c_src = crate_dir.join("tests/smoke_ffi.c");
    let include_dir = crate_dir.join("include");

    // Find the target directory (where libleyline_fs lives)
    let target_dir = crate_dir.join("../../target/debug");

    let out_binary = tmp.path().join("smoke_ffi");

    // Compile the C test — platform-specific linker flags
    let mut cc_args = vec![
        "-o".to_string(),
        out_binary.to_str().unwrap().to_string(),
        c_src.to_str().unwrap().to_string(),
        format!("-I{}", include_dir.display()),
        format!("-L{}", target_dir.display()),
        "-lleyline_fs".to_string(),
    ];
    if cfg!(target_os = "macos") {
        cc_args.extend([
            "-framework".to_string(),
            "Security".to_string(),
            "-framework".to_string(),
            "CoreFoundation".to_string(),
            "-lresolv".to_string(),
        ]);
    } else {
        // Linux: link pthread, dl, m (pulled in by rusqlite/fuser)
        cc_args.extend([
            "-lpthread".to_string(),
            "-ldl".to_string(),
            "-lm".to_string(),
        ]);
    }
    let compile = Command::new("cc").args(&cc_args).output()?;

    if !compile.status.success() {
        let stderr = String::from_utf8_lossy(&compile.stderr);
        panic!("C compilation failed:\n{}", stderr);
    }

    // Run the C test
    let run = Command::new(out_binary.to_str().unwrap())
        .arg(ctrl_path.to_str().unwrap())
        .env("DYLD_LIBRARY_PATH", target_dir.to_str().unwrap())
        .output()?;

    let stdout = String::from_utf8_lossy(&run.stdout);
    let stderr = String::from_utf8_lossy(&run.stderr);

    if !run.status.success() {
        panic!(
            "C smoke test failed (exit {}):\nstdout:\n{}\nstderr:\n{}",
            run.status, stdout, stderr
        );
    }

    println!("{}", stdout);
    assert!(
        stdout.contains("PASS"),
        "expected PASS in output:\n{}",
        stdout
    );

    Ok(())
}
